py/objdeque.c

```diff
@@ -135,6 +183,74 @@ STATIC mp_obj_t deque_popleft(mp_obj_t self_in) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(deque_popleft_obj, deque_popleft);
 
+STATIC mp_obj_t deque_pop(mp_obj_t self_in) {
+    mp_obj_deque_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (self->i_get == self->i_put) {
+        mp_raise_msg(&mp_type_IndexError, MP_ERROR_TEXT("empty"));
+    }
+
+    if (self->i_put == 0) {
+        self->i_put = self->alloc - 1;
+    } else {
+        self->i_put--;
+    }
+
+    mp_obj_t ret = self->items[self->i_put];
+
+    return ret;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(deque_pop_obj, deque_pop);
+
+STATIC mp_obj_t deque_getiter(mp_obj_t o_in, mp_obj_iter_buf_t *iter_buf) {
+    mp_obj_deque_t *deque = MP_OBJ_TO_PTR(o_in);
+    return mp_obj_new_deque_iterator(o_in, deque->i_get, iter_buf);
+}
+
+STATIC void mp_obj_deque_store(mp_obj_t self_in, mp_obj_t index, mp_obj_t value) {
+    mp_obj_deque_t *self = MP_OBJ_TO_PTR(self_in);
+    size_t offset = mp_get_index(self->base.type, deque_len(self), index, false);
+    size_t i = (self->i_get + offset) % self->alloc;
+    self->items[i] = value;
+}
+
+STATIC mp_obj_t deque_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value) {
+    if (value == MP_OBJ_NULL) {
+        // delete not supported, but must match CPython error message
+        #if MICROPY_PY_BUILTINS_SLICE
+        if (mp_obj_is_type(index, &mp_type_slice)) {
+            mp_raise_TypeError(MP_ERROR_TEXT("sequence index must be integer, not slice"));
+        }
+        #endif
+        #if MICROPY_ERROR_REPORTING <= MICROPY_ERROR_REPORTING_TERSE
+        mp_raise_NotImplementedError(MP_ERROR_TEXT("object doesn't support item deletion"));
+        #else
+        mp_raise_msg_varg(&mp_type_NotImplementedError,
+            MP_ERROR_TEXT("'%s' object doesn't support item deletion"), mp_obj_get_type_str(self_in));
+        #endif
+    } else if (value == MP_OBJ_SENTINEL) {
+        // load
+        mp_obj_deque_t *self = MP_OBJ_TO_PTR(self_in);
+        #if MICROPY_PY_BUILTINS_SLICE
+        if (mp_obj_is_type(index, &mp_type_slice)) {
+            mp_raise_TypeError(MP_ERROR_TEXT("sequence index must be integer, not slice"));
+        }
+        #endif
+        size_t index_offset = mp_get_index(self->base.type, deque_len(self), index, false);
+        size_t index_val = (self->i_get + index_offset) % self->alloc;
+        return self->items[index_val];
+    } else {
+        // store into deque
+        #if MICROPY_PY_BUILTINS_SLICE
+        if (mp_obj_is_type(index, &mp_type_slice)) {
+            mp_raise_TypeError(MP_ERROR_TEXT("sequence index must be integer, not slice"));
+        }
+        #endif
+        mp_obj_deque_store(self_in, index, value);
```

Comment:
If you inline this function here then load and store have a lot of code in common and can be factored to reduce code size.